cmd/kubeadm/app/cmd/phases/kubelet.go

```diff
@@ -111,72 +94,73 @@ var (
 		`)
 )
 
-// NewCmdKubelet returns command for `kubeadm phase kubelet`
-func NewCmdKubelet() *cobra.Command {
-	cmd := &cobra.Command{
-		Use:   "kubelet",
-		Short: "Commands related to handling the kubelet.",
-		Long:  cmdutil.MacroCommandLongDescription,
-	}
-
-	cmd.AddCommand(NewCmdKubeletConfig())
-	cmd.AddCommand(NewCmdKubeletWriteEnvFile())
-	return cmd
+// kubeletStartData defines the behavior that a runtime data struct passed to the kubelet start phase
+// should have. Please note that we are using an interface in order to make this phase reusable in different workflows
+// (and thus with different runtime data struct, all of them requested to be compliant to this interface)
+type kubeletStartData interface {
+	Cfg() *kubeadmapi.InitConfiguration
+	DryRun() bool
+	KubeletDir() string
 }
 
-// NewCmdKubeletWriteEnvFile calls cobra.Command for writing the dynamic kubelet env file based on a InitConfiguration or JoinConfiguration object
-func NewCmdKubeletWriteEnvFile() *cobra.Command {
-	var cfgPath string
-
-	cmd := &cobra.Command{
-		Use:     "write-env-file",
-		Short:   "Writes an environment file with runtime flags for the kubelet.",
-		Long:    kubeletWriteEnvFileLongDesc,
-		Example: kubeletWriteEnvFileExample,
-		Run: func(cmd *cobra.Command, args []string) {
-			if len(cfgPath) == 0 {
-				kubeadmutil.CheckErr(fmt.Errorf("The --config flag is mandatory"))
-			}
-
-			err := RunKubeletWriteEnvFile(cfgPath)
-			kubeadmutil.CheckErr(err)
-		},
+// NewKubeletStartPhase creates a kubeadm workflow phase that start kubelet on a node.
+func NewKubeletStartPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:    "kubelet-start",
+		Short:   "Writes kubelet settings and (re)starts kubelet",
+		Long:    "Writes a file with KubeletConfiguration and an environment file with node specific kubelet settings, and then (re)starts kubelet.",
+		Example: kubeletStartPhaseExample,
+		Run:     runKubeletStart,
 	}
-
-	options.AddConfigFlag(cmd.Flags(), &cfgPath)
-	return cmd
 }
 
-// RunKubeletWriteEnvFile is the function that is run when "kubeadm phase kubelet write-env-file" is executed
-func RunKubeletWriteEnvFile(cfgPath string) error {
-	internalcfg, err := configutil.AnyConfigFileAndDefaultsToInternal(cfgPath)
-	if err != nil {
-		return err
+// runKubeletStart executes start kubelet logic.
+func runKubeletStart(c workflow.RunData) error {
+	data, ok := c.(kubeletStartData)
```

Comment:
what's the reason behind using the `kubeletStartData` interface instead of casting it to a `*initData`?

Reply:
It's local to this phase.